libbpf-rs/src/btf/mod.rs

```diff
@@ -0,0 +1,149 @@
+//! Parse and introspect btf information, from files or loaded objects.
+//!
+//! To find a specific type you can use one of 3 methods
+//!
+//! - [Btf::type_by_name]
+//! - [Btf::type_by_id]
+//! - [Btf::type_by_kind]
+//!
+//! All of these are generic over `K`, which is any type that can be created from a [`BtfType`],
+//! for all of these methods, not finding any type by the passed parameter or finding a type of
+//! another [`BtfKind`] will result in a [`None`] being returned (or filtered out in the case of
+//! [`Btf::type_by_kind`]). If you want to get a type independently of the kind, just make sure `K`
+//! binds to [`BtfType`].
+
+use std::ffi::CStr;
+use std::ffi::CString;
+use std::marker::PhantomData;
+use std::mem::size_of;
+use std::os::raw::c_void;
+use std::os::unix::prelude::AsRawFd;
+use std::os::unix::prelude::FromRawFd;
+use std::os::unix::prelude::OsStrExt;
+use std::os::unix::prelude::OwnedFd;
+use std::path::Path;
+use std::ptr::NonNull;
+
+use crate::libbpf_sys;
+use crate::util::create_bpf_entity_checked;
+use crate::util::create_bpf_entity_checked_opt;
+use crate::util::parse_ret_i32;
+use crate::Error;
+use crate::Result;
+
+/// The btf information of a bpf object.
+///
+/// The lifetime bound protects against this object outliving its source. This can happen when it
+/// was derived from an [`Object`](super::Object), which owns the data this structs points too. When
+/// instead the [`Btf::from_path`] method is used, the lifetime will be `'static` since it doesn't
+/// borrow from anything.
+#[derive(Debug)]
+pub struct Btf<'source> {
+    ptr: NonNull<libbpf_sys::btf>,
+    needs_drop: bool,
+    _marker: PhantomData<&'source ()>,
+}
+
+impl Btf<'static> {
+    /// Load the btf information from an ELF file.
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let path = {
+            let mut v = path.as_ref().as_os_str().as_bytes().to_vec();
+            v.push(0);
+            CString::from_vec_with_nul(v).map_err(|_| {
+                Error::InvalidInput(format!("invalid path {:?}, has null bytes", path.as_ref()))
+            })?
+        };
+        let ptr = create_bpf_entity_checked(|| unsafe {
+            libbpf_sys::btf__parse_elf(path.as_ptr(), std::ptr::null_mut())
```

Comment:
from [this](https://github.com/libbpf/libbpf/blob/f8cd00f61302ca4d8645e007acf7f546dc323b33/src/btf.c#L1048-L1056) code in libbpf  (and reading the code of the function they delegate to) I concluded it was okay to pass null to that parameter. Since that function lacked documentation I chose to pass null and not provide a way for the caller to override that parameter for now, since I didn't understand it. In the future a `from_path_with_ext` method could be added to support passing a value there.